patches/tModLoader/Terraria/NPC.cs.patch

```diff
@@ -706,6 +706,24 @@
  			if (Main.npc[newNPC].type == 1 && Main.player[num7].RollLuck(180) == 0)
  				Main.npc[newNPC].SetDefaults(-4);
  
+@@ -62468,7 +_,7 @@
+ 				ChatHelper.BroadcastChatMessage(NetworkText.FromKey("Announcement.HasAwoken", Main.npc[num].GetTypeNetName()), new Color(175, 75, 255));
+ 		}
+ 
+-		public static int NewNPC(int X, int Y, int Type, int Start = 0, float ai0 = 0f, float ai1 = 0f, float ai2 = 0f, float ai3 = 0f, int Target = 255) {
++		public static int NewNPC(int X, int Y, int Type, int Start = 0, float ai0 = 0f, float ai1 = 0f, float ai2 = 0f, float ai3 = 0f, int Target = 255, INPCSource source = null) {
+ 			if (Main.getGoodWorld) {
+ 				if (Type == 46)
+ 					Type = 614;
+@@ -62517,6 +_,8 @@
+ 						ChatHelper.BroadcastChatMessage(NetworkText.FromKey("Announcement.HasAwoken", Main.npc[num].GetTypeNetName()), new Color(175, 75, 255));
+ 				}
+ 
++				NPCLoader.OnSpawn(Main.npc[num], source);
```

Comment:
I just found one annoying vanilla code detail: The code that spawns NPCs does something special for NPCs with negative `netID`, which are just variants of other NPCs, where it does something like:
```cs
int index = NPC.NewNPC(spawnTileX * 16 + 8, spawnTileY * 16, NPCID.BlueSlime); // ID 1
Main.npc[index].SetDefaults(NPCID.GreenSlime); // ID -3
```
Now, because `NPC.SetDefaults` is being called again after `NPC.NewNPC`, it will probably wipe whatever the `NPCLoader.OnSpawn` call may have done. 
On a similar note, a problem with tModLoader currently is that the call to `NPCLoader.SetDefaults` is at a place were it doesn't receive the correct `netID` for negative NPCs, due to how default values of negative `netID` NPCs are set. Idk if fixing that may be within the scope of this PR though.